feature_engine/timeseries/forecasting/expanding_window_features.py

```diff
@@ -0,0 +1,232 @@
+
+
+@Substitution(
+    variables=_variables_numerical_docstring,
+    missing_values=_missing_values_docstring,
+    drop_original=_drop_original_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit=_fit_not_learn_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class ExpandingWindowFeatures(BaseForecast):
+    """
+    ExpandingWindowFeatures adds new features to a dataframe based on window
+    operations. Expanding window operations are operations that perform an
+    aggregation over an expanding window of all past values relative to the
+    value of interest. An expanding window feature is, in other words, a feature
+    created after computing statistics (e.g., mean, min, max, etc.) using a
+    window over all the past data. For example, the mean value of all months
+    prior to the month of the interest is an expanding window feature.
+
+    ExpandingWindowFeatures uses the pandas' functions `expanding()`, `agg` and
+    `shift`.  With `expanding`, it creates expanding windows. With `agg` it
+    applies multiple functions within those windows. With 'shift()' it allocates
+    the values to the correct rows.
+
+    For supported aggregation functions, see Expanding Window
+    `Functions <https://pandas.pydata.org/docs/reference/window.html>`_.
+
+    To be compatible with ExpandingWindowFeatures, the dataframe's index must
+    have unique values and no NaN.
+
+    ExpandingWindowFeatures works only with numerical variables. You can pass a
+    list of variables to use as input for the expanding window. Alternatively,
+    ExpandingWindowFeatures will automatically select all numerical variables
+    in the training set.
+
+    More details in the :ref:`User Guide <window_features>`.
+
+    Parameters
+    ----------
+    {variables}
+
+    min_periods: int, default None.
+        Minimum number of observations in window required to have a value;
+        otherwise, result is np.nan. See parameter `min_periods` in the pandas
+        `expanding` documentation for more details.
+
+    functions: str, list of str, default = 'mean'
+        The functions to apply within the window. Valid functions can be found
+        `here <https://pandas.pydata.org/docs/reference/window.html>`_.
+
+    sort_index: bool, default=True
+        Whether to order the index of the dataframe before creating window
+        feature.
+
+    {missing_values}
+
+    {drop_original}
+
+    Attributes
+    ----------
+    variables_:
+        The group of variables that will be used to create the window features.
+
+    {feature_names_in_}
+
+    {n_features_in_}
+
+    Methods
+    -------
+    {fit}
+
+    transform:
+        Add window features.
+
+    {fit_transform}
+
+    See Also
+    --------
+    pandas.expanding
+    pandas.aggregate
+    pandas.shift
+    """
+
+    def __init__(
+        self,
+        variables: None | int | str | list[str | int] = None,
```

Comment:
I understand the concern around consistency. Here are some thoughts. Using | was introduced with Python 3.10 under PEP 604. Using type annotations in this way reduces complexity (e.g.,  no need to import Union or List from the typing module and it is also easier to read). In other code bases (e.g., Statsmodels) the type annotations are not forced to stick to the old style for consistency.

Given this, would you still want to make the trade off for consistency? Longer term, would feature-engine accept contributions which make use of PEP 604? If so it may be easier to accept contributions which use it rather than a bunch of PRs to migrate type annotations.

Reply:
Makes sense. Thank you for pointing to PEP604. Lets use | going forward.